File: apps/desktop/src/services/commands/decks.ts
import { invoke } from "@tauri-apps/api/core";

export interface DiscoverDeckDto {
  id: number;
  slug: string;
  title: string;
  description: string | null;
  level: string | null;
  wordCount: number;
  tags: string[];
  sampleWords?: string[];
  packName: string;
  packSlug: string;
  installed: boolean;
}

export interface DiscoverDecksDto {
  decks: DiscoverDeckDto[];
  total: number;
}

export interface LibraryDeckDto {
  id: number;
  slug: string;
  title: string;
  description: string | null;
  level: string | null;
  wordCount: number;
  tags: string[];
  sampleWords?: string[];
  packName: string;
  packSlug: string;
  installedAt: string;
  masteredCount: number;
  dueCount: number;
  accuracy: number;
  lastStudied: string | null;
  progress: number;
}

export interface LibraryDecksDto {
  decks: LibraryDeckDto[];
  total: number;
}

export interface DeckPreviewWordDto {
  id: number;
  headword: string;
  partOfSpeech: string | null;
  level: string | null;
  definitionEn: string | null;
  definitionVi: string | null;
  example: string | null;
  dueState: "Due today" | "Learning" | "Mastered" | "New";
}

export interface DeckDetailProgressDto {
  masteredCount: number;
  dueCount: number;
  accuracy: number;
  lastStudied: string | null;
  progress: number;
}

export interface DeckDetailDto {
  id: number;
  slug: string;
  title: string;
  description: string | null;
  level: string | null;
  wordCount: number;
  tags: string[];
  packName: string;
  packSlug: string;
  banner: string | null;
  installed: boolean;
  installedAt: string | null;
  progress: DeckDetailProgressDto;
  words: DeckPreviewWordDto[];
}

export function listDiscoverDecks(): Promise<DiscoverDecksDto> {
  return invoke<DiscoverDecksDto>("list_discover_decks");
}

export function listLibraryDecks(): Promise<LibraryDecksDto> {
  return invoke<LibraryDecksDto>("list_library_decks");
}

export function getDeckDetail(deckId: number): Promise<DeckDetailDto> {
  return invoke<DeckDetailDto>("get_deck_detail", { deckId });
}

export function installDeck(deckId: number): Promise<void> {
  return invoke<void>("install_deck", { deckId });
}

export function uninstallDeck(deckId: number): Promise<void> {
  return invoke<void>("uninstall_deck", { deckId });
}

File: apps/desktop/src-tauri/src/dto/decks.rs
use serde::Serialize;

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct DeckSummaryDto {
    pub id: i64,
    pub slug: String,
    pub name: String,
    pub description: Option<String>,
    pub word_count: i64,
    pub difficulty: Option<String>,
    /// Raw JSON array string stored in the database, e.g. `["A1","beginner"]`.
    pub tags: Option<String>,
    pub pack_name: String,
    pub pack_slug: String,
}

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct SeededDecksDto {
    pub decks: Vec<DeckSummaryDto>,
    pub total: usize,
}

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct DiscoverDeckDto {
    pub id: i64,
    pub slug: String,
    pub title: String,
    pub description: Option<String>,
    pub level: Option<String>,
    pub word_count: i64,
    pub tags: Vec<String>,
    pub sample_words: Vec<String>,
    pub pack_name: String,
    pub pack_slug: String,
    pub installed: bool,
}

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct DiscoverDecksDto {
    pub decks: Vec<DiscoverDeckDto>,
    pub total: usize,
}

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct LibraryDeckDto {
    pub id: i64,
    pub slug: String,
    pub title: String,
    pub description: Option<String>,
    pub level: Option<String>,
    pub word_count: i64,
    pub tags: Vec<String>,
    pub sample_words: Vec<String>,
    pub pack_name: String,
    pub pack_slug: String,
    pub installed_at: String,
    pub mastered_count: i64,
    pub due_count: i64,
    pub accuracy: i64,
    pub last_studied: Option<String>,
    pub progress: i64,
}

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct LibraryDecksDto {
    pub decks: Vec<LibraryDeckDto>,
    pub total: usize,
}

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct DeckPreviewWordDto {
    pub id: i64,
    pub headword: String,
    pub part_of_speech: Option<String>,
    pub level: Option<String>,
    pub definition_en: Option<String>,
    pub definition_vi: Option<String>,
    pub example: Option<String>,
    pub due_state: String,
}

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct DeckDetailProgressDto {
    pub mastered_count: i64,
    pub due_count: i64,
    pub accuracy: i64,
    pub last_studied: Option<String>,
    pub progress: i64,
}

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct DeckDetailDto {
    pub id: i64,
    pub slug: String,
    pub title: String,
    pub description: Option<String>,
    pub level: Option<String>,
    pub word_count: i64,
    pub tags: Vec<String>,
    pub pack_name: String,
    pub pack_slug: String,
    pub banner: Option<String>,
    pub installed: bool,
    pub installed_at: Option<String>,
    pub progress: DeckDetailProgressDto,
    pub words: Vec<DeckPreviewWordDto>,
}
